torch/_refs/__init__.py

```diff
@@ -811,140 +480,114 @@ def float_power(
 )
 
 # TODO: add docstring
 ne = _make_elementwise_binary_reference(
-    prims.ne, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.ALWAYS_BOOL
+    prims.ne, type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.ALWAYS_BOOL
 )
 
 # TODO: add docstring
 nextafter = _make_elementwise_binary_reference(
-    prims.nextafter, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
+    prims.nextafter, type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
 )
 
 # TODO: add docstring
 pow = _make_elementwise_binary_reference(
-    prims.pow, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.BOOL_TO_LONG
+    prims.pow, type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.BOOL_TO_LONG
 )
 
 # TODO: add docstring
-# Sub is implemented specially because it has an alpha argument and
-#   is decomposed into multiple prims.
 # TODO: consider refactoring this with add impl
+# sub has its own implementation because it has an alpha argument
 @register_decomposition(torch.ops.aten.sub)
+@out_wrapper
+@elementwise_type_promotion_wrapper(
+    type_promoting_args=("a", "b"),
+    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.OP_MATH,
+)
 def sub(
-    a: Union[Tensor, NumberType],
-    b: Union[Tensor, NumberType],
+    a: Union[TensorLikeType, NumberType],
+    b: Union[TensorLikeType, NumberType],
     *,
     alpha: Optional[NumberType] = None,
-    out: Optional[Tensor] = None
 ):
     """
-    Reference implementation of torch.sub
+    Reference implementation of torch.add
     """
-
-    # Type checks
-    assert isinstance(a, (TensorLike, Number))
-    assert isinstance(b, (TensorLike, Number))
-    assert out is None or isinstance(out, TensorLike)
-    assert alpha is None or isinstance(alpha, Number)
-
-    # Special-cases Number x Number case
-    if isinstance(a, Number) and isinstance(b, Number):
-        a, b = utils.wrap_scalars(a, b)
-
-    computation_dtype, result_dtype = _elementwise_dtypes(
-        a, b, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.OP_MATH
-    )
-    a, b = _convert_dtype(a, b, dtype=computation_dtype)
-
-    a, b = _unwrap_cpu_scalars(a, b)
-
-    a, b = broadcast(a, b)
+    a, b = _maybe_broadcast(a, b)
 
     if alpha is not None:
-        alpha_promotion_type = utils.dtype_to_type(computation_dtype)
-        assert utils.is_lesser_type(type(alpha), alpha_promotion_type) or (
-            computation_dtype is torch.bool and type(alpha) is int
-        )
-        b = prims.mul(b, alpha_promotion_type(alpha))
-
-    result = prims.sub(a, b)
-
-    (result,) = _convert_dtype(result, dtype=result_dtype)
-
-    if out is not None:
-        out = _maybe_resize_out(out, result.shape)
-        return copy_to(out, result, allow_cross_device=False)  # type: ignore[arg-type]
+        dtype = a.dtype if isinstance(a, TensorLike) else b.dtype  # type: ignore[union-attr]
+        python_type = utils.dtype_to_type(dtype)
+        if not utils.is_weakly_lesser_type(type(alpha), python_type):
```

Comment:
why do we need both `is_weakly_lesser_type` and `can_safely_cast_to` that do the same thing?

Reply:
There's probably an opportunity to unify them but one currently works on types and the other on dtypes.

Reply:
Right, but you have maps to convert between the two.